### ISSUE TITLE
HttpClient throwing null reference exception on sending request

### ISSUE THREAD

**Issue by AeddGinvael** [bluemagic123/tModLoader#3073]
### Version

Latest 1.4-stable

### OS

Linux

### Platform

Steam

### This bug affects

Mod capability as a Modder

### Description

`HttpClient.SendAsync` always failing with Null reference exception on latest stable version `v2022.09.47.13`

### Log File

[server.log](https://github.com/tModLoader/tModLoader/files/9911888/server.log)

### Steps to reproduce

1. Make a request, for example, like this
```
var client = new HttpClient();
var req = new HttpRequestMessage
{
    Content = new StringContent("<random json>", Encoding.UTF8, "application/json"),
    Method = HttpMethod.Post,
    RequestUri = new Uri("<uri>"), 
};

var response = await client.SendAsync(req);
```
2. line `var response = await _client.SendAsync(req);` always throwing `System.NullReferenceException: Object reference not set to an instance of an object`

### Expected Behavior

`client.SendAsync` (or any other method of http client) not throwing null reference exception

### Actual Behavior

Any request made by `HttpClient` is throwing null reference exception

### Reproduction frequency

Reproduced at all times on latest stable version

### Additional Information

 I am using [docker container](https://github.com/EdwardJXLi/tmodloader-docker) for hosting my tmodloader server on linux. 
 Everything was fine until i updated tmodloader version from v2022.08.54.6 to v2022.9.47.13
 Also, I tried disabling mods other than mine, but the error still appeared. Everything works locally without errors.

**Comment by JavidPack** [comment]
Have you tested on windows? does the error not happen there?

**Comment by AeddGinvael** [comment]
Yes, I've checked. There was no problem with that

**Comment by JavidPack** [comment]
K, I was able to replicate on wsl linux, but haven't found the root cause yet.

**Comment by xNarnia** [comment]
Hello! I've taken time to debug this issue and I am posting my findings here.

Below is a server.log with logs of orig and various properties within it. 
As well, I have logged all of the parameters being supplied to orig on the return. 
`orig(self, request, proxyUri, async, doRequestAuth, isProxyConnect, cancellationToken);`

It errors out as soon as I try to log **self** with a NullReferenceException
[server.log](https://github.com/tModLoader/tModLoader/files/10189842/server.log)

It is also worth noting that this error occurs on both 1.4.3.6 and 1.4.4, so updating Terraria doesn't fix the issue

### Additional Information

**The payload that is being downloaded**
![image](https://user-images.githubusercontent.com/7970069/206579252-3d642048-7459-42ab-9893-c1823fff9be0.png)

**Successful execution of HttpRequest**
![image](https://user-images.githubusercontent.com/7970069/206579622-26c025fb-c354-43ce-a0ce-ea023e744bf4.png)

**The erroring code commented out**
![image](https://user-images.githubusercontent.com/7970069/206579244-9eae95cb-b8bc-4496-889a-cd5576f6047f.png)

**Link to the Erroring Code**
https://github.com/tModLoader/tModLoader/blob/c51c8f832ccd52311f81a628ffbf4df3bb775bbe/patches/tModLoader/Terraria/ModLoader/Engine/LoggingHooks.cs#L75-L96

**Comment by xNarnia** [comment]
After talking with the MonoMod guys, they seem to have pinpointed a spot where the issue could be occurring. 

Edit: A fix is now being PR'ed in

Discord Thread
https://discord.com/channels/295566538981769216/488446768326115328/1051991820734828554

Screenshots
![image](https://user-images.githubusercontent.com/7970069/207190107-5c35a9ad-ec82-4ada-854b-b1a2612ab764.png)
![image](https://user-images.githubusercontent.com/7970069/207190035-7ec21212-a2bb-4a60-aa18-4c3c91f0df98.png)

**Comment by phonetic112** [comment]
I still have this issue:

```
[13:37:20.507] [.NET ThreadPool Worker/DEBUG] [tML]: Web Request: https://api.openai.com/v1/chat/completions
[13:37:20.508] [.NET ThreadPool Worker/WARN] [tML]: Silently Caught Exception: 
System.NullReferenceException: Object reference not set to an instance of an object.
```

On Linux also.

**Comment by Hanprogramer** [comment]
Still not working
![image](https://user-images.githubusercontent.com/30620423/230766650-16caf1c1-1b17-408b-a6a7-858f44d439f0.png)

**Comment by Gunging** [comment]
_tModLoader server 1.4.3.6+2022.09.47.29|stable|Stable|6c8f8d7b415992bca8b9dc5147efc3a6a6dda189|5249794107657932047 built 02/01/2023 00:42_

Could this be related?
```
[00:33:05.018] [.NET ThreadPool Worker/DEBUG] [tML]: Web Request: https://discord.com/api/v10/gateway/bot
[00:33:05.023] [.NET ThreadPool Worker/WARN] [tML]: Silently Caught Exception: 
System.NullReferenceException: Object reference not set to an instance of an object.
   at DMD<System.Net.Http.HttpConnectionPoolManager::SendAsyncCore>(HttpConnectionPoolManager this, HttpRequestMessage request, Uri proxyUri, Boolean async, Boolean doRequestAuth, Boolean isProxyConnect, CancellationToken cancellationToken)
   at Trampoline<System.Net.Http.HttpConnectionPoolManager::SendAsyncCore>?29441404(Object , HttpRequestMessage , Uri , Boolean , Boolean , Boolean , CancellationToken )
   
   at Terraria.ModLoader.Engine.LoggingHooks.<>c.<HookWebRequests>b__12_0(orig_SendAsyncCore orig, Object self, HttpRequestMessage request, Uri proxyUri, Boolean async, Boolean doRequestAuth, Boolean isProxyConnect, CancellationToken cancellationToken) in tModLoader\Terraria\ModLoader\Engine\LoggingHooks.cs:line 84
   
   at Hook<System.Net.Http.HttpConnectionPoolManager::SendAsyncCore>?58204539(HttpConnectionPoolManager , HttpRequestMessage , Uri , Boolean , Boolean , Boolean , CancellationToken )
   at System.Net.Http.HttpConnectionPoolManager.SendAsync(HttpRequestMessage request, Boolean async, Boolean doRequestAuth, CancellationToken cancellationToken)
   at System.Net.Http.HttpMessageHandlerStage.SendAsync(HttpRequestMessage request, CancellationToken cancellationToken)
   at System.Net.Http.DiagnosticsHandler.SendAsync(HttpRequestMessage request, Boolean async, CancellationToken cancellationToken)
   at System.Net.Http.RedirectHandler.SendAsync(HttpRequestMessage request, Boolean async, CancellationToken cancellationToken)
   at System.Runtime.CompilerServices.AsyncMethodBuilderCore.Start[TStateMachine](TStateMachine& stateMachine)
   at System.Net.Http.RedirectHandler.SendAsync(HttpRequestMessage request, Boolean async, CancellationToken cancellationToken)
   at System.Net.Http.DecompressionHandler.SendAsync(HttpRequestMessage request, Boolean async, CancellationToken cancellationToken)
   at System.Runtime.CompilerServices.AsyncMethodBuilderCore.Start[TStateMachine](TStateMachine& stateMachine)
   at System.Net.Http.DecompressionHandler.SendAsync(HttpRequestMessage request, Boolean async, CancellationToken cancellationToken)
   at System.Net.Http.SocketsHttpHandler.SendAsync(HttpRequestMessage request, CancellationToken cancellationToken)
   at System.Net.Http.HttpClientHandler.SendAsync(HttpRequestMessage request, CancellationToken cancellationToken)
   at System.Net.Http.HttpMessageInvoker.SendAsync(HttpRequestMessage request, CancellationToken cancellationToken)
   at System.Net.Http.HttpClient.<SendAsync>g__Core|83_0(HttpRequestMessage request, HttpCompletionOption completionOption, CancellationTokenSource cts, Boolean disposeCts, CancellationTokenSource pendingRequestsCts, CancellationToken originalCancellationToken)
   at System.Runtime.CompilerServices.AsyncMethodBuilderCore.Start[TStateMachine](TStateMachine& stateMachine)
   at System.Net.Http.HttpClient.<SendAsync>g__Core|83_0(HttpRequestMessage request, HttpCompletionOption completionOption, CancellationTokenSource cts, Boolean disposeCts, CancellationTokenSource pendingRequestsCts, CancellationToken originalCancellationToken)
   at System.Net.Http.HttpClient.SendAsync(HttpRequestMessage request, HttpCompletionOption completionOption, CancellationToken cancellationToken)
   at System.Net.Http.HttpClient.SendAsync(HttpRequestMessage request, CancellationToken cancellationToken)
   at Discord.Net.Rest.DefaultRestClient.SendInternalAsync(HttpRequestMessage request, CancellationToken cancelToken, Boolean headerOnly)
   at System.Runtime.CompilerServices.AsyncMethodBuilderCore.Start[TStateMachine](TStateMachine& stateMachine)
   at Discord.Net.Rest.DefaultRestClient.SendInternalAsync(HttpRequestMessage request, CancellationToken cancelToken, Boolean headerOnly)
   at Discord.Net.Rest.DefaultRestClient.SendAsync(String method, String endpoint, CancellationToken cancelToken, Boolean headerOnly, String reason, IEnumerable`1 requestHeaders)
   at System.Runtime.CompilerServices.AsyncMethodBuilderCore.Start[TStateMachine](TStateMachine& stateMachine)
   at Discord.Net.Rest.DefaultRestClient.SendAsync(String method, String endpoint, CancellationToken cancelToken, Boolean headerOnly, String reason, IEnumerable`1 requestHeaders)
   at Discord.Net.Queue.RestRequest.SendAsync()
   at System.Runtime.CompilerServices.AsyncMethodBuilderCore.Start[TStateMachine](TStateMachine& stateMachine)
   at Discord.Net.Queue.RestRequest.SendAsync()
   at Discord.Net.Queue.RequestBucket.SendAsync(RestRequest request)
   at System.Runtime.CompilerServices.AsyncMethodBuilderCore.Start[TStateMachine](TStateMachine& stateMachine)
   at Discord.Net.Queue.RequestBucket.SendAsync(RestRequest request)
   at Discord.Net.Queue.RequestQueue.SendAsync(RestRequest request)
   at System.Runtime.CompilerServices.AsyncMethodBuilderCore.Start[TStateMachine](TStateMachine& stateMachine)
   at Discord.Net.Queue.RequestQueue.SendAsync(RestRequest request)
   at Discord.API.DiscordRestApiClient.SendInternalAsync(String method, String endpoint, RestRequest request)
   at System.Runtime.CompilerServices.AsyncMethodBuilderCore.Start[TStateMachine](TStateMachine& stateMachine)
   at Discord.API.DiscordRestApiClient.SendInternalAsync(String method, String endpoint, RestRequest request)
   at Discord.API.DiscordRestApiClient.SendAsync[TResponse](String method, String endpoint, BucketId bucketId, ClientBucketType clientBucket, RequestOptions options)
   at System.Runtime.CompilerServices.AsyncMethodBuilderCore.Start[TStateMachine](TStateMachine& stateMachine)
   at Discord.API.DiscordRestApiClient.GetBotGatewayAsync(RequestOptions options)
   at System.Runtime.CompilerServices.AsyncMethodBuilderCore.Start[TStateMachine](TStateMachine& stateMachine)
   at Discord.API.DiscordRestApiClient.GetBotGatewayAsync(RequestOptions options)
   at Discord.API.DiscordSocketApiClient.ConnectInternalAsync()
   at System.Runtime.CompilerServices.AsyncMethodBuilderCore.Start[TStateMachine](TStateMachine& stateMachine)
   at Discord.API.DiscordSocketApiClient.ConnectInternalAsync()
   at Discord.API.DiscordSocketApiClient.ConnectAsync()
   at System.Runtime.CompilerServices.AsyncMethodBuilderCore.Start[TStateMachine](TStateMachine& stateMachine)
   at Discord.API.DiscordSocketApiClient.ConnectAsync()
   at Discord.WebSocket.DiscordSocketClient.OnConnectingAsync()
   at System.Runtime.CompilerServices.AsyncMethodBuilderCore.Start[TStateMachine](TStateMachine& stateMachine)
   at Discord.WebSocket.DiscordSocketClient.OnConnectingAsync()
   at Discord.ConnectionManager.ConnectAsync(CancellationTokenSource reconnectCancelToken)
   at System.Runtime.CompilerServices.AsyncMethodBuilderCore.Start[TStateMachine](TStateMachine& stateMachine)
   at Discord.ConnectionManager.ConnectAsync(CancellationTokenSource reconnectCancelToken)
   at Discord.ConnectionManager.<>c__DisplayClass29_0.<<StartAsync>b__0>d.MoveNext()
   at System.Runtime.CompilerServices.AsyncMethodBuilderCore.Start[TStateMachine](TStateMachine& stateMachine)
   at Discord.ConnectionManager.<>c__DisplayClass29_0.<StartAsync>b__0()
   at System.Threading.Tasks.Task`1.InnerInvoke()
   at System.Threading.ExecutionContext.RunFromThreadPoolDispatchLoop(Thread threadPoolThread, ExecutionContext executionContext, ContextCallback callback, Object state)
   at System.Threading.Tasks.Task.ExecuteWithThreadLocal(Task& currentTaskSlot, Thread threadPoolThread)
   at System.Threading.Tasks.Task.ExecuteFromThreadPool(Thread threadPoolThread)
   at System.Threading.ThreadPoolWorkQueue.Dispatch()
   at System.Threading.PortableThreadPool.WorkerThread.WorkerThreadStart()
   at System.Threading.Thread.StartCallback()
   ```
   
   EDIT 1: Highlighted the Terraria.ModLoader.Engine.LoggingHooks method between new lines
   
   EDIT 2: Included tModLoader version

**Comment by Andr3wD** [comment]
I have this issue, but with WebSockets. After testing, it looks like this issue is fixed on TML 1.4.4-preview, but it's likely that existing mods will need to be migrated from TML 1.4.3 for the fix.

**Comment by JavidPack** [comment]
I believe the changes to tmodloader in 1.4.4 that fixed this involved a large breaking change to a dependency we use, so it is probably not likely that we can fix this issue on 1.4.3, since it would break existing mods. 1.4.4 is progressing smoothly, I hope we can get it stable soon so the majority of users can benefit from the changes and fixes.

**Comment by Chicken-Bones** [comment]
Closing as fixed in 1.4.4, we unfortunately won't be able to backport this to 1.4.3

**Comment by richerp** [comment]
I meet the same question
